### PR TITLE
BatchNorm splitting into ops (2nd try)

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: src/ngraph/op/batch_norm.cpp [hunk 1]
```diff
@@ -21,51 +21,98 @@
...
+    set_output_type(0, et, bn_input_shape);
+}
 
+void ngraph::op::BatchNormTraining::validate_and_infer_types()
+{
+    auto bn_input_shape = get_input_shape(INPUT);
+    BatchNormBase::validate_and_infer_types();
+    auto in_size = get_input_size();
+    auto& et = get_input_element_type(INPUT);
     NODE_VALIDATION_ASSERT(this, in_size == 3 || in_size == 5)
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: I think we should split the 3-input and 5-input variants into separate ops.

Comment (diyessi) [reply src/ngraph/op/batch_norm.cpp]: Yes. One is used in training, the other in inference.

File: src/ngraph/op/batch_norm.cpp [hunk 2]
```diff
@@ -21,51 +21,90 @@
...
-    : Op("BatchNorm", check_single_output_args({gamma, beta, input}))
-    , m_bn_input_shape(input->get_shape())
+ngraph::op::BatchNormBase::BatchNormBase(const std::string& node_type,
+                                         double eps,
+                                         const NodeVector& args)
+    : Op(node_type, check_single_output_args(args))
     , m_epsilon(eps)
-    , m_training(true)
 {
     constructor_validate_and_infer_types();
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: I think we should not call this in a `Base` class. @diyessi

File: src/ngraph/op/batch_norm.cpp [hunk 3]
```diff
@@ -21,51 +21,90 @@
...
+    constructor_validate_and_infer_types();
+}
 
+void ngraph::op::BatchNormInference::validate_and_infer_types()
+{
+    auto bn_input_shape = get_input_shape(INPUT);
+    BatchNormBase::validate_and_infer_types();
+    auto in_size = get_input_size();
     auto& et = get_input_element_type(INPUT);
+    NODE_VALIDATION_ASSERT(this, in_size == 5)
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: I think this check is redundant, because `Node`'s constructor will set the input count.

File: src/ngraph/op/batch_norm.cpp [hunk 4]
```diff
@@ -21,51 +21,90 @@
...
+    set_output_type(0, et, bn_input_shape);
+}
+
+void ngraph::op::BatchNormTraining::validate_and_infer_types()
+{
+    auto bn_input_shape = get_input_shape(INPUT);
+    BatchNormBase::validate_and_infer_types();
     auto in_size = get_input_size();
+    auto& et = get_input_element_type(INPUT);
+    NODE_VALIDATION_ASSERT(this, in_size == 3)
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: Ditto---I think this is redundant.

File: src/ngraph/op/batch_norm.cpp [hunk 5]
```diff
@@ -198,57 +201,47 @@ void ngraph::op::BatchNormBackprop::validate_and_infer_types()
...
-            throw ngraph_error("GetOutputElement for mean is missing");
-        };
-        if (!var)
-        {
-            throw ngraph_error("GetOutputElement for variance is missing");
-        }
-    }
-    else // BatchNorm Training with global stats
+        throw ngraph_error("GetOutputElement for mean is missing");
+    };
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: stray semicolon

**Comment by Krovatkin** [comment]
@postrational Could one of your guys help me with ONNX failures?

File: src/ngraph/op/batch_norm.cpp (hunk 3 above)
Comment (Krovatkin) [reply src/ngraph/op/batch_norm.cpp]: fixed

File: src/ngraph/op/batch_norm.cpp (hunk 4 above)
Comment (Krovatkin) [reply src/ngraph/op/batch_norm.cpp]: fixed

File: src/ngraph/op/batch_norm.cpp (hunk 5 above)
Comment (Krovatkin) [reply src/ngraph/op/batch_norm.cpp]: removed

**Comment by postrational** [comment]
ONNX test are passing now, but we'll need to merge a change in nGraph-ONNX together with this (https://github.com/NervanaSystems/ngraph-onnx/pull/212).
Let me know when this is ready to merge and I'll merge both.

File: src/ngraph/pass/core_fusion.cpp [hunk 6]
```diff
@@ -202,15 +202,15 @@ void pass::CoreFusion::construct_folded_batch_norm()
     auto beta = std::make_shared<pattern::op::Label>(element::f32, beta_shape);
     double eps = 0.001;
     auto shape_r = Shape{1, 2, 2, 2};
-    auto bn = std::make_shared<op::BatchNorm>(eps, gamma, beta, pconv, mean, var);
+    auto bn = std::make_shared<op::BatchNormInference>(eps, gamma, beta, pconv, mean, var);
```
Comment (pruthviIntel) [inline src/ngraph/pass/core_fusion.cpp]: does the batch norm fording only applies to `Inference` Op?. I think  we can apply the below transformation during training as well.
```
        // new weights = old weights * gamma / sqrt(variance + epsilon)
        // new biases = -mean * gamma / sqrt(variance + epsilon) + beta
```

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 7]
```diff
@@ -809,46 +809,62 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_batch_norm_relu_global_sta
...
+            bn_relu = std::make_shared<op::BatchNormInferenceRelu>(bn_inference->get_eps_value(),
+                                                                   pattern_map[gamma],
+                                                                   pattern_map[beta],
+                                                                   pattern_map[input],
+                                                                   pattern_map[mean],
+                                                                   pattern_map[var]);
+        }
+        else if (auto bn_training = std::dynamic_pointer_cast<op::BatchNormTraining>(bn_match))
+        {
+            bn_relu = std::make_shared<op::BatchNormTrainingRelu>(bn_training->get_eps_value(),
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: this is little confusing, why are we replacing with `batchnorm relu` fused Op with global stats. It efficient  for bn training we compute the `global stats` within mkldnn kernel.
